### PR TITLE
refactors Journalist Interface for semantic HTML5/ARIA markup

### LINKED ISSUE #5987: semantic (HTML5/ARIA) page structure in Journalist Interface
# Description

[Accessibility Lab recommendations](https://docs.google.com/document/d/1jZlKco2YgOrrI4PmTkG_Q8lKuV5VxCds/edit#) include:

- `h1`...`hn` hierarchy
- HTML5 section elements
- ARIA roles ("landmarks"), labels, and descriptions

Our findings include:

- moving decorative `img` elements to CSS styles on semantic elements
- Keep in mind per @eloquence in <https://github.com/freedomofpress/securedrop/issues/5972#issuecomment-947317684>:
    > One thing we may want to look into for the next round of changes is how to minimize the need to re-translate strings - e.g., should we generate our own translations in cases where only HTML elements or attributes change (can we do so in the main repo and then import these into Weblate?).

## User Research Evidence

[WCAG 1.3.1](https://www.w3.org/TR/UNDERSTANDING-WCAG20/content-structure-separation-programmatic.html) per #5972.

## Progress

To invert the way things are usually done :-):

1. [x] **First pass:**  Refactor for semantic markup (test in developer tools and screen-readers) and keep `journalist.js` in sync
    - [x] decorative `img`/`hr`/etc. elements: comment out as `FIXME`s to move to CSS styles on semantic elements 
    - [x] linters pass
    - [x] tests pass
2. [x] **Second pass:**  Audit in validators and accessibility tools ([axe](https://www.deque.com/axe/), [WAVE](https://wave.webaim.org/)), e.g.:
    - `h1`...`hn` hierarchy
    - HTML5 section elements
    - ARIA roles ("landmarks"), labels, and descriptions
3. [x] Accessibility-focused review of draft PR before continuing further
4. [x] Review string changes for translation per <https://github.com/freedomofpress/securedrop/issues/5972#issuecomment-947317684> 
5. [x] **Third pass:**  CSS and images broken in first pass (test visually in Tor Browser)
    - [x] `FIXME`s
        - [x] Revert 6d4602da288b398dea3680dfb9faba67cf11e85a
    - [x] tests pass
6. [x] Write and run through test plan in #6165

| Path | First Pass | Second Pass | Third Pass |
| --- | --- | --- | --- |
| `securedrop/journalist_templates` | ✓ | ✓ | ✓ |
| `├──_confirmation_modal.html` | ✓ | ✓| ✓ |
| `├──_source_row.html` | ✓ | ✓ | ✓ |
| `├──_sources_confirmation_final_modal.html` | ✓ | ✓ | ✓ |
| `├──_sources_confirmation_modal.html` | ✓ | ✓ | ✓ |
| `├──account_edit_hotp_secret.html` | ✓ | ✓ | ✓ |
| `├──account_new_two_factor.html` | ✓ | ✓ | ✓ |
| `├──admin.html` | ✓ | ✓ | ✓ |
| `├──admin_add_user.html` | ✓ | ✓ | ✓ |
| `├──admin_edit_hotp_secret.html` | ✓ | ✓ | ✓ |
| `├──admin_new_user_two_factor.html` | ✓ | ✓ | ✓ |
| `├──base.html` | ✓ | ✓ | ✓ |
| `├──col.html` | ✓ | ✓ | ✓ |
| `├──config.html` | ✓ | ✓ | ✓ |
| `├──delete.html` | ✓ | ✓ | ✓ |
| `├──edit_account.html` | ✓ | ✓ | ✓ |
| `├──error.html` | ✓ | ✓ | ✓ |
| `├──flashed.html` | ✓ | ✓ | ✓ |
| `├──index.html` | ✓ | ✓ | ✓ |
| `├──js-strings.html` | ✓ | ✓ | ✓ |
| `├──locales.html` | ✓ | ✓ | ✓ |
| `├──login.html` | ✓ | ✓ | ✓ |
| `└──preferences_saved_flash.html` | ✓ | ✓ | ✓ |

## Specific cases

- [x] Address review feedback in https://github.com/freedomofpress/securedrop/pull/6041#discussion_r673883016
- [x] Refactor `*modal.html` as `dialog`s à la Source Interface's `.{confirm,hidden}-prompt`

### `add_user.html`
- [x] `h1` should not be `.visually-hidden`

### `col.html`
- [x] Listing of a source collection could be a semantic `table`
- [x] `securedrop/journalist_templates/col.html`: `p.breadcrumbs` is a set of unstructured inline elements; could be `ul` styled with `li:{before,after}` selectors
- [x] `div dialog`s are still being read (unlike those in `index.html`)

### `index.html`
- [x] List of sources could be a semantic `table`
    - [x] `td.filename` should wrap more elegantly (or not at all)

### PR DESCRIPTION
**Pull request by cfm**
## Status

* Accessibility changes approved by @SaptakS 
* Ready for final maintainer's review
    * See "Checklist" section for pre-merge steps

## Description of Changes

Closes #5987 by:

1. [x] refactoring all `securedrop/journalist_templates/*.html` for semantic markup;
2. [x] adding ARIA annotations where HTML5 semantics are insufficient; and
3. [x] fixing SASS/CSS in light of (1)–(2), especially to move decorative images out of the DOM and accessibility trees.

## Testing

Click through the Journalist Interface and inspect in terms of:

1. ***Markup,* by inspecting and validating HTML.** This need not take place in Tor Browser; I used the [Web Developer extension](https://chrispederick.com/work/web-developer/) in Chrome.
2. ***Accessibility in assistive tech,* by listening to the interface as announced from Tor Browser by a screen-reader.**  It should sound and "feel" conspicuously more fluent.  (I also audited pages in [WebAIM's Wave Evaluation Tool](https://chrome.google.com/webstore/detail/wave-evaluation-tool/jbbplnpkjmmeebjpijfedlgcdilocofh) and [Deque's axe DevTools](https://chrome.google.com/webstore/detail/axe-devtools-web-accessib/lhdoppojpmngadmnindnejefpokejbdd), but these will yield mixed results until the conclusion of this accessibility push.)
3. ***Visual appearance,* by comparing the interface against <https://demo-journalist.securedrop.org>.**  Changes are intended to be "pixel-perfect" except where there were preexisting idiosyncrasies that more-semantic markup tended to make naturally more consistent (or where noted in individual commits).  Of course, suggestions for fine-tuning are welcome!

Steps (2)–(3) should be performed in Tor Browser at both "standard" and "safest" security levels.

Finally:

4. Click through the Source Interface and compare against <https://demo-source.securedrop.org> for any regressions introduced by shared Sass.

## Deployment

No special considerations.

### After approval and before merge

1. [x] Reformat `securedrop/journalist_templates/*.html`
2. [x] Update as many translations as possible per <https://github.com/freedomofpress/securedrop/issues/5972#issuecomment-947317684>; test in Weblate sandbox  →  see <https://github.com/freedomofpress/securedrop/pull/6165#issuecomment-1016857219>
3. Approver will squash-merge ~(especially fix-ups) to one commit per pass~

### If you made changes to the server application code:

- [x] Linting (`make lint`) and tests (`make test`) pass in the development container

### CODE REVIEW

**Comment by cfm** [comment]
@SaptakS, while this pull request is officially still a draft, I wanted to request your accessibility review of the markup refactoring I've proposed here.  Per discussion with the SecureDrop team, once I've revised this draft based on your feedback, I'll make the CSS changes necessary for pixel-perfect consistency with `develop`, for final review by @zenmonkeykstop.

Note that one test (`functional/test_submit_and_retrieve_message.py::TestSubmitAndRetrieveMessage::test_submit_and_retrieve_happy_path`) is still flaking.  Once I've come up with a fix, I'll append that commit to this branch to avoid disrupting your review, Saptak, and then squash it in prior to merge.

**Comment by cfm** [comment]
Confirmed that `functional/test_submit_and_retrieve_message.py::TestSubmitAndRetrieveMessage::test_submit_and_retrieve_happy_path` is indeed a flake, with local passes reproduced at last in CI as of <https://app.circleci.com/pipelines/github/freedomofpress/securedrop/3208/workflows/70299cd8-95de-48df-b79b-c7a4cf6bb764>.

**Comment by cfm** [comment]
The flake, most demonstrably seen in <https://app.circleci.com/pipelines/github/freedomofpress/securedrop/3208/workflows/dd8b4348-b646-4cbb-8ac1-af76e1173290/jobs/58091/tests#failed-test-0>, looks like:

```
self = <tests.functional.test_submit_and_retrieve_message.TestSubmitAndRetrieveMessage object at 0x7ff49d42ebb0>

    def test_submit_and_retrieve_happy_path(self):
        self._source_visits_source_homepage()
        self._source_chooses_to_submit_documents()
        self._source_continues_to_submit_page()
        self._source_submits_a_message()
        self._source_logs_out()
        self.switch_to_firefox_driver()
        self._journalist_logs_in()
        self._journalist_checks_messages()
>       self._journalist_downloads_message()

/home/circleci/project/securedrop/tests/functional/test_submit_and_retrieve_message.py:20: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/home/circleci/project/securedrop/tests/functional/journalist_navigation_steps.py:798: in _journalist_downloads_message
    self._journalist_selects_the_first_source()
/home/circleci/project/securedrop/tests/functional/journalist_navigation_steps.py:790: in _journalist_selects_the_first_source
    self.driver.find_element_by_css_selector("#un-starred-source-link-1").click()
[...]
E       selenium.common.exceptions.ElementNotInteractableException: Message: Element <a id="un-starred-source-link-1" class="code-name unread" href="/col/E7DV7R4U3FALMXE3N2THUA7DQQUPAUUKFOENZPO7FGEGRELBC3XQM22KMMR53TOJFAN3XILJ6FQTZGRDG6DDNRXXF66JXBXAYMJXABA="> could not be scrolled into view

/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/selenium/webdriver/remote/errorhandler.py:242: ElementNotInteractableException
```

This behavior is not reproducible in manual testing (in Chrome, Firefox, or Tor Browser) and in 20546031a64d647926bff7f011c4a92ab7240600 persists even if `JournalistNavigationStepsMixin._journalist_selects_the_first_source()` uses `FunctionalTest.safe_click_by_id()` (as it probably should).  I've tinkered quite a bit with the test logic, to no avail, and will now poke at the markup to see if I can understand what Selenium is complaining about.  Per eb9f35da08cabd0001bac29120c12bb21db7cb48, it doesn't *appear* to be the new overlaid `dialog` elements.

File: securedrop/journalist_templates/_source_row.html
Comment (SaptakS) [inline securedrop/journalist_templates/_source_row.html]: I don't see the need for an aria-label here, since the inner html already contains the time element with the same text content.

File: securedrop/journalist_templates/account_edit_hotp_secret.html
Comment (SaptakS) [inline securedrop/journalist_templates/account_edit_hotp_secret.html]: There needs to be an `id="otp_secret"` other the for attribute in label doesn't really make sense. https://developer.mozilla.org/en-US/docs/Web/HTML/Element/label#attr-for

File: securedrop/journalist_templates/account_edit_hotp_secret.html
Comment (SaptakS) [inline securedrop/journalist_templates/account_edit_hotp_secret.html]: Adding an `aria-label` here may make sense in cake a more verbose text is added. The idea is if links and buttons have similar text throughout the website, e.g. Continue/cancel might be used in different places, it's better to have an `aria-label` that says continue to what. Something like `Continue to change secret` or something better. Just to convert from all caps doesn't make a lot of sense to me.

File: securedrop/journalist_templates/account_new_two_factor.html
Comment (SaptakS) [inline securedrop/journalist_templates/account_new_two_factor.html]: 👍🏽

File: securedrop/journalist_templates/account_new_two_factor.html
Comment (SaptakS) [inline securedrop/journalist_templates/account_new_two_factor.html]: Same as mentioned above since `Submit` is also pretty common text that appears repeatedly.

File: securedrop/journalist_templates/admin.html
Comment (SaptakS) [inline securedrop/journalist_templates/admin.html]: I feel this entire section can just be an `<a>` tag. Clicking the `Add User` takes the user to a separate page, so I think that qualifies as a link. Don't see the point of a form with button, unless, I am missing something.

File: securedrop/journalist_templates/admin.html
Comment (SaptakS) [inline securedrop/journalist_templates/admin.html]: Why convert it into a section instead of just an `<a>` tag? You can add the text inside h2 as the `aria-label` for the anchor tag since it's a bit verbose.

File: securedrop/journalist_templates/admin_add_user.html
Comment (SaptakS) [inline securedrop/journalist_templates/admin_add_user.html:33]: Don't think this `<div>` is needed. There are two consecutive `<div>`s instead of just one.

File: securedrop/journalist_templates/admin_add_user.html
Comment (SaptakS) [inline securedrop/journalist_templates/admin_add_user.html:40]: Same as above.

File: securedrop/journalist_templates/admin_edit_hotp_secret.html
Comment (SaptakS) [inline securedrop/journalist_templates/admin_edit_hotp_secret.html]: Needs `id` attribute for `for` to work.

**Review by SaptakS — COMMENTED** [review]
This looks awesome! Left few comments that I think should be addressed. Some of the comments are applicable in multiple places and templates (e.g., missing id attribute and aria-label on buttons comments ). The comments are related to only HTML semantics, since I think the styling is still under progress and there are some "Fix me" sections.

File: securedrop/journalist_templates/_source_row.html
Comment (cfm) [reply securedrop/journalist_templates/_source_row.html]: I agree in theory!  In practice, at least my screen-reader (VoiceOver on macOS) reads the `datetime` attribute instead of the inner HTML, unless overridden with this `aria-label`.

File: securedrop/journalist_templates/account_edit_hotp_secret.html
Comment (cfm) [reply securedrop/journalist_templates/account_edit_hotp_secret.html]: Absolutely; this was sloppy on my part.  Fixed in f54157a73.

File: securedrop/journalist_templates/account_edit_hotp_secret.html
Comment (cfm) [reply securedrop/journalist_templates/account_edit_hotp_secret.html]: Thanks.  I've expanded a number of `aria-label`s in 39954528f.  In other places, the change of capitalization remains to prevent screen-readers from mistaken all-caps words for initialisms.

File: securedrop/journalist_templates/admin.html
Comment (cfm) [reply securedrop/journalist_templates/admin.html]: Great catch; thank you.  I'm not sure why 54834e22309e3c8dfaead2a07a650a9af87e7ca1 added this `button` rather than an `a.btn`, which is what 2be57f94ea50683c5a7a3d63c67f25a3097219bf subsequently did for the **Instance Config** button.  Fixed in f4d25065d6d3a74bbbbae7e264f91173389de3f9.

File: securedrop/journalist_templates/admin_add_user.html
Comment (cfm) [reply securedrop/journalist_templates/admin_add_user.html:33]: Thanks!  I'd misread this as necessary for the flexbox layout.  Fixed in cbc3b8ace.

File: securedrop/journalist_templates/admin_edit_hotp_secret.html
Comment (cfm) [reply securedrop/journalist_templates/admin_edit_hotp_secret.html]: Also fixed in f54157a73.

File: securedrop/journalist_templates/admin.html
Comment (cfm) [reply securedrop/journalist_templates/admin.html]: I appreciate this simplification—done in f4d25065d6d3a74bbbbae7e264f91173389de3f9.

**Comment by cfm** [comment]
Thanks, @SaptakS!  I've been able to resolve all but one of your suggestions right away.  I appreciate your careful eye on these changes!  On now to CSS polishing.

**Comment by cfm** [comment]
This pull request's _substantive_ changes are complete with the conclusion of CSS polishing in 07f9b57f.  Remaining work per #5987:

1. get tests passing;
2. do another round of functional and visual testing (Tor Browser) and validation;
3. reformat `*.html` à la 8f2b5031f9ae1f55586a7eeab52cf4503538ee02.

**Review by zenmonkeykstop — CHANGES_REQUESTED** [review]
- [x] HTML validation (in Tor Browser, via the "validate local html" option of the Web Developer plugin looks good.
- [x] Page is readable via Tails' screen reader
- [x] No visual discrepancies that stood out vs the demo site
- [ ] *FAIL* The Source Interface seems mostly unchanged except for one point:
  - the **Get started** and **Log in** buttons are truncated, and text is not centered in them. (this is relatively major)
  - (for other reviewers, the lock icon on the generate page has been removed in develop, and the "forgot your codename" link in flash messages is on its own line now. These are not issues in this PR, just the demo site lagging behind the develop branch)

I'll poke about a bit to see what's affecting the button padding on the SI, but this looks OK to merge  once that's fixed.

**Comment by cfm** [comment]
<https://github.com/freedomofpress/securedrop/pull/6165#pullrequestreview-851972869>:

> [ ] _FAIL_ The Source Interface seems mostly unchanged except for one point:
> * the **Get started** and **Log in** buttons are truncated, and text is not centered in them. (this is relatively major)

Ack; thanks, @zenmonkeykstop.  I'd fixed this at some point but must have lost track of it in a stash or rebase.  Should be a quick fix to reconstruct (or just dig out of my local reflog).

**Comment by cfm** [comment]
@zenmonkeykstop, this should be fixed in cbfc08f.  Once you've reviewed and approved, I'll start in (either today or Monday) on the "After Approval and Before Merge" checklist above.

**Comment by zenmonkeykstop** [comment]
Confirmed that the SI fix looks good. If you wanna knock off the first two items on that list and leave the squash merge to the approver, I can take a final pass once they're in place.

File: securedrop/journalist_templates/base.html
Comment (legoktm) [inline securedrop/journalist_templates/base.html]: nit: I think this shouldn't be in the div and just directly right before `</body>`

File: securedrop/journalist_templates/edit_account.html
Comment (legoktm) [inline securedrop/journalist_templates/edit_account.html]: Not sure if it's in scope, but these labels seem like they should be shown for all users instead of relying solely on the placeholder text.

File: securedrop/journalist_templates/base.html
Comment (cfm) [reply securedrop/journalist_templates/base.html]: Good call, @legoktm; thanks.  Fixup in 340938556.

File: securedrop/journalist_templates/edit_account.html
Comment (cfm) [reply securedrop/journalist_templates/edit_account.html]: I have mixed feelings about the label/placeholder usage across SecureDrop forms too.  In the Journalist Interface:

* labels are visible everywhere fields require explication (e.g., `admin_add_user.html`);
* labels are visible and augmented with placeholders where fields have logical relations (e.g., label "Is using a Yubikey" + placeholder "HOTP Secret"); and
* labels are visually hidden in favor of placeholders, but still present in the accessibility tree, where fields have logical relations familiar enough to be left implicit (e.g., `login.html`'s three fields and the two fields you've cited in `edit_account.html`).

In this case in `edit_account.html`, it seems to me that the two fields here:

1. visually imitate the three fields of the login form;
2. have explanatory text nearby in— <https://github.com/freedomofpress/securedrop/blob/cbfc08f73c56afa970fbdadf8ecbffe4d6d2e7e3/securedrop/journalist_templates/edit_account.html#L65>—and
3. still benefit from their placeholder text.

So, for kicks, here is c9eb580a74e9b066d71724a5845c37d1aa6342d7, where these fields are grouped in a `fieldset` with an explanatory `legend`, otherwise keeping the `login.html`-like field layout.  I'll be curious to know whether you and @zenmonkeykstop like this or think it's more trouble than it's worth.  :-)

**Comment by cfm** [comment]
Re: the pre-merge step:

> Update as many translations as possible per <https://github.com/freedomofpress/securedrop/issues/5972#issuecomment-947317684>; test in Weblate sandbox

Unfortunately, after running `make translate` on this branch, the only developer-friendly translation drift I see is (e.g., for `ar`):

```patch
-msgid "Powered by <em>SecureDrop {version}</em>."
+#, fuzzy
+#| msgid "Powered by <em>SecureDrop {version}</em>."
+msgid "Powered by <b>SecureDrop {version}</b>."
 msgstr "من <em>SecureDrop {version} </em>."
```

I'll be happy to do an `s/em/b` pass in Weblate as usual at the start of the translation period.  In the current translation workflow, however, I don't think it's worth trying to be clever about updating the `.po`s directly here for a single HTML fix.

File: securedrop/journalist_templates/edit_account.html
Comment (legoktm) [reply securedrop/journalist_templates/edit_account.html]: I like the fieldset, it's a minimal change but IMO improves the usability of it.

**Review by legoktm — DISMISSED** [review]
The changes made in response to my feedback LGTM. I did a final skim as well, but not any full testing (I assume Kev already did or will do that)

**Comment by cfm** [comment]
Thanks, @legoktm!  I've rebased out the `WIP` commit message and a few `fixup`s.  With that, this is ready for @zenmonkeykstop's final review and merge.

**Comment by zenmonkeykstop** [comment]
Thanks all, taking a final spin through now.

**Review by zenmonkeykstop — APPROVED** [review]
Took a quick spin through validation and screen-reader steps again, no worries on that score.
 Source Interface glitch is resolved, looks consistent overall.

LGTM - thanks @cfm for the work and @SaptakS and @legotm for review!

**Pull request merged**
